### PR TITLE
GCB-134: Added a mapping between the harvest models and view models for the main chart

### PR DESCRIPTION
**Pull request by karolkozub**

### CODE REVIEW

File: GrandCentralBoard/Widgets/Harvest/HarvestWidget.swift [hunk 1]
```diff
@@ -36,20 +36,13 @@ final class HarvestWidget : Widget {
...
-
-        let model = AreaBarChartViewModel(mainChart: mainChart, componentCharts: [history, history, history, history, history], horizontalAxisStops: 20, horizontalAxisLabelText: "people billing:", centerText: nil)
-
-        widgetView.render(model)
-
-        // Why another render? To check if view renders the view model after the previous one.
-        // TODO: Remove!
-        // widgetView.render(AreaBarChartViewModel.emptyViewModel(header: "HARVEST BURN REPORT", subheader: "Monday 28.03.2016"))
+        case .Failure(_):
+            break
```
Comment (ochococo) [inline GrandCentralBoard/Widgets/Harvest/HarvestWidget.swift]: Call `widgetView.failure()` here.

File: GrandCentralBoard/Widgets/Harvest/View/AreaBarChartComponentViewModel+FromDailyBillingStats.swift [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+//  Created by Karol Kozub on 2016-04-18.
+//  Copyright © 2016 Oktawian Chojnacki. All rights reserved.
+//
+
+import UIKit
+
+
+extension AreaBarChartComponentViewModel {
+    static func viewModelFromDailyBillingStats(dailyBillingStats: DailyBillingStats) -> AreaBarChartComponentViewModel? {
+        guard dailyBillingStats.groups.count == 3 else {
```
Comment (ochococo) [inline GrandCentralBoard/Widgets/Harvest/View/AreaBarChartComponentViewModel+FromDailyBillingStats.swift]: Do we really need all 3? Anything between 1 and 3 is okay IMO.

File: GrandCentralBoard/Widgets/Harvest/View/AreaBarChartComponentViewModel+FromDailyBillingStats.swift [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        let colors = [UIColor.lipstick(), UIColor.aquaMarine(), UIColor.almostWhite()]
+
+        let items = indexes.flatMap { index -> AreaBarItemViewModel? in
+            let group = groups[index]
+            let color = colors[index]
+            let width = widthForGroup(group, groups: groups)
+            let height = heightForGroup(group, groups: groups)
+            let valueLabelMode = valueLabelModeForGroup(group, groups: groups)
+
+            if (width > 0) {
```
Comment (ochococo) [inline GrandCentralBoard/Widgets/Harvest/View/AreaBarChartComponentViewModel+FromDailyBillingStats.swift]: `guard` ?

File: GrandCentralBoard/Widgets/Harvest/View/AreaBarChartComponentViewModel+FromDailyBillingStats.swift [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    private static func countLabelTextForDailyBillingStats(dailyBillingStats: DailyBillingStats) -> String {
+        let totalCount = totalCountForGroups(dailyBillingStats.groups)
+
+        return String(totalCount)
+    }
+
+    private static func headerTextForDailyBillingStats(dailyBillingStats: DailyBillingStats) -> String {
+        return "HARVEST BURN REPORT"
```
Comment (ochococo) [inline GrandCentralBoard/Widgets/Harvest/View/AreaBarChartComponentViewModel+FromDailyBillingStats.swift]: `NSLocalizedString`?

File: GrandCentralBoard/Widgets/Harvest/View/AreaBarChartViewModel+FromBillingStats.swift [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+import GrandCentralBoardCore
+
+
+extension AreaBarChartViewModel {
+    static func viewModelFromBillingStats(billingStats: [DailyBillingStats]) -> AreaBarChartViewModel {
+        let componentViewModels = billingStats.flatMap({ dailyBillingStats in
+            return AreaBarChartComponentViewModel.viewModelFromDailyBillingStats(dailyBillingStats)
+        })
+
+        if (componentViewModels.count < 1) {
```
Comment (ochococo) [inline GrandCentralBoard/Widgets/Harvest/View/AreaBarChartViewModel+FromBillingStats.swift]: `componentViewModels.isEmpty`

Comment (ochococo) [inline GrandCentralBoard/Widgets/Harvest/View/AreaBarChartViewModel+FromBillingStats.swift]: drop the parenthesis

Comment (ochococo) [inline GrandCentralBoard/Widgets/Harvest/View/AreaBarChartViewModel+FromBillingStats.swift]: or `guard let mainViewModel = componentViewModels.first else { return AreaBarChartViewModel.emptyViewModel() }`

File: GrandCentralBoard/Widgets/Harvest/View/AreaBarChartViewModel+FromBillingStats.swift [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+            return AreaBarChartViewModel.emptyViewModel()
+        }
+
+        let mainViewModel = componentViewModels.first!
```
Comment (ochococo) [inline GrandCentralBoard/Widgets/Harvest/View/AreaBarChartViewModel+FromBillingStats.swift]: Use `guard` above and avoid this call.

File: GrandCentralBoardTests/AreaBarChartComponentViewModel+Equatable.swift [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+import Foundation
+@testable import GrandCentralBoard
+
+
+extension AreaBarItemValueLabelDisplayMode : Equatable {
+}
+
+
+func ==(lhs: AreaBarItemValueLabelDisplayMode, rhs: AreaBarItemValueLabelDisplayMode) -> Bool {
```
Comment (ochococo) [inline GrandCentralBoardTests/AreaBarChartComponentViewModel+Equatable.swift]: Nice.

File: GrandCentralBoardTests/AreaBarChartComponentViewModelFromDailyBillingStatsTests.swift [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+//
+//  Created by Karol Kozub on 2016-04-18.
+//  Copyright © 2016 Oktawian Chojnacki. All rights reserved.
+//
+
+import XCTest
+@testable import GrandCentralBoard
+
+
+class AreaBarChartComponentViewModelFromDailyBillingStatsTests: XCTestCase {
```
Comment (ochococo) [inline GrandCentralBoardTests/AreaBarChartComponentViewModelFromDailyBillingStatsTests.swift]: 👏 TESTS!

File: configuration.json [hunk 9]
```diff
@@ -3,16 +3,9 @@
...
-                             "http://i2.cdn.turner.com/cnnnext/dam/assets/150324154025-14-internet-cats-restricted-super-169.jpeg",
-                             "http://www.cats.org.uk/uploads/images/pages/photo_latest14.jpg",
-                             "https://pbs.twimg.com/profile_images/378800000532546226/dbe5f0727b69487016ffd67a6689e75a.jpeg"
-                             ]
-                    }
-                },
-                {"name": "harvest", "settings": {"account": "The name used in harvest URLs, e.g. `gcbtest` from http://gcbtest.harvestapp.com", "refreshToken": "", "clientID": "", "clientSecret": "", "numberOfDays": 6, "refreshInterval": 3600}}
+                {"name":"watch", "settings":  {"timeZone":"Europe/Warsaw", "calendar": "http://localhost:8000/calendar.json"} },
+                {"name": "harvest", "settings": {"account": "gcbtest", "refreshToken": "kifUPmZ_Ha_QcwtrZIFT7H4J0NVoHBXpotvO4bvvrO7Jz50ou6RcMDCVivuX0DXaA3BrMOy_gHBQaRHTjJsLFw", "clientID": "BE8GLh27kjQ7MIswYuD57A", "clientSecret": "12aTwDPIko5NKjoTMo_uczmnvts_umAOZqrEFW44AqlCpsZn6IhCG-fsKoAEXO2clE3ayNxv5hTmMtd46dPruA", "numberOfDays": 6, "refreshInterval": 3600}}
```
Comment (ochococo) [inline configuration.json]: Remove the token 🐛

File: GrandCentralBoard/Widgets/Harvest/View/AreaBarChartComponentViewModel+FromDailyBillingStats.swift [hunk 10]
```diff
@@ -48,15 +48,43 @@ extension AreaBarChartComponentViewModel {
...
+
     private static func heightForGroup(group: BillingStatsGroup, groups: [BillingStatsGroup]) -> CGFloat {
-        let maxHours = 8.0
-        let hours = min(group.averageHours, maxHours)
+        let hours = max(minHours, min(maxHours, group.averageHours))
 
         return CGFloat(hours / maxHours)
     }
 
+    private static func doubleText(value: Double) -> String {
```
Comment (ochococo) [inline GrandCentralBoard/Widgets/Harvest/View/AreaBarChartComponentViewModel+FromDailyBillingStats.swift]: `extension Double` and with meaningful method name.

`elegantString` or `stringWithoutPointZero`?

File: GrandCentralBoard/Widgets/Harvest/View/AreaBarChartComponentViewModel+FromDailyBillingStats.swift [hunk 11]
```diff
@@ -48,15 +48,43 @@ extension AreaBarChartComponentViewModel {
...
+        if (Double(Int(value)) == value) {
+            return String(Int(value))
+
+        } else {
+            return String(value)
```
Comment (ochococo) [inline GrandCentralBoard/Widgets/Harvest/View/AreaBarChartComponentViewModel+FromDailyBillingStats.swift]: format

File: GrandCentralBoard/Widgets/Harvest/View/AreaBarChartComponentViewModel+FromDailyBillingStats.swift [hunk 12]
```diff
@@ -48,15 +48,43 @@ extension AreaBarChartComponentViewModel {
...
+        }
+    }
+
+    private static func hoursText(hours: Double) -> String {
+        if (hours > maxHours) {
+            return String(format: "More than %@!", doubleText(maxHours))
```
Comment (ochococo) [inline GrandCentralBoard/Widgets/Harvest/View/AreaBarChartComponentViewModel+FromDailyBillingStats.swift]: remove capital letter

File: GrandCentralBoard/Widgets/Harvest/View/AreaBarChartComponentViewModel+FromDailyBillingStats.swift [hunk 13]
```diff
@@ -48,15 +48,43 @@ extension AreaBarChartComponentViewModel {
...
+
+        } else if (hours < minHours) {
+            return String(format: "Less than %@!", doubleText(minHours))
```
Comment (ochococo) [inline GrandCentralBoard/Widgets/Harvest/View/AreaBarChartComponentViewModel+FromDailyBillingStats.swift]: remove capital letter

**Comment by karolkozub** [comment]
I've applied the suggested changes. @ochococo

**Comment by karolkozub** [comment]
I've rebased the branch. @ochococo

File: GrandCentralBoard/Widgets/Harvest/BillingStatsFetcher.swift [hunk 14]
```diff
@@ -54,8 +54,8 @@ final class BillingStatsFetcher {
 
 
 extension NSDate {
-    static func lastNDays(numberOfDays: Int) -> [NSDate] {
-        return (0..<numberOfDays).map({ index in
+    static func aNumberOfPreviousDays(numberOfDays: Int) -> [NSDate] {
```
Comment (ochococo) [inline GrandCentralBoard/Widgets/Harvest/BillingStatsFetcher.swift]: change nae as discussed

**Comment by ochococo** [comment]
LGTM 👍

File: GrandCentralBoard/Widgets/Harvest/HarvestWidget.swift [hunk 15]
```diff
@@ -42,7 +42,7 @@ final class HarvestWidget : Widget {
             widgetView.render(model)
 
         case .Failure(_):
```
Comment (ochococo) [inline GrandCentralBoard/Widgets/Harvest/HarvestWidget.swift]: `.Failure` is enough.

File: GrandCentralBoard/Widgets/Harvest/View/AreaBarChartComponentViewModel+FromDailyBillingStats.swift [hunk 16]
```diff
@@ -137,3 +128,18 @@ extension NSDate {
...
+
+extension Double {
+    var formattedString: String {
+        let intValue = Int(self)
+        let isInteger = Double(intValue) == self
+
+        if isInteger {
+            return String(intValue)
+
+        } else {
```
Comment (ochococo) [inline GrandCentralBoard/Widgets/Harvest/View/AreaBarChartComponentViewModel+FromDailyBillingStats.swift]: `else` is not needed.

File: GrandCentralBoard.xcodeproj/project.pbxproj [hunk 17]
```diff
@@ -1247,6 +1250,7 @@
 				1A87495E1C385264006E58C6 /* RemoteImageSource.swift in Sources */,
 				F9C88F3E1CBFC72100E81A14 /* AccessTokenFetcher.swift in Sources */,
 				52FB50621CBE862D0053DE8B /* EventsSource.swift in Sources */,
+				F928B82E1CC52F77006D0332 /* String+Localized.swift in Sources */,
```
Comment (ochococo) [inline GrandCentralBoard.xcodeproj/project.pbxproj]: `String+Localized` -> `String`

File: GrandCentralBoard/Widgets/Harvest/View/AreaBarChartComponentViewModel+FromDailyBillingStats.swift [hunk 18]
```diff
@@ -104,7 +95,7 @@ extension AreaBarChartComponentViewModel {
 
     private static func headerTextForDailyBillingStats(dailyBillingStats: DailyBillingStats, isMainChart: Bool) -> String {
```
Comment (ochococo) [inline GrandCentralBoard/Widgets/Harvest/View/AreaBarChartComponentViewModel+FromDailyBillingStats.swift]: `isMainChart` -> `forMainChart` ?

File: GrandCentralBoard/Widgets/Harvest/View/AreaBarChartComponentViewModel+FromDailyBillingStats.swift [hunk 19]
```diff
@@ -104,7 +95,7 @@ extension AreaBarChartComponentViewModel {
...
         if (isMainChart) {
-            return "HARVEST BURN REPORT"
+            return "HARVEST BURN REPORT".localized
 
         } else {
```
Comment (ochococo) [inline GrandCentralBoard/Widgets/Harvest/View/AreaBarChartComponentViewModel+FromDailyBillingStats.swift]: `} else {` is not needed.

**Comment by ochococo** [comment]
No more comments. 🎾 @karolkozub

File: GrandCentralBoard/Widgets/Harvest/View/AreaBarChartComponentViewModel+FromDailyBillingStats.swift (hunk 16 above)
Comment (karolkozub) [inline GrandCentralBoard/Widgets/Harvest/View/AreaBarChartComponentViewModel+FromDailyBillingStats.swift]: I prefer to have `return` statements on the same indentation level if possible

File: GrandCentralBoard.xcodeproj/project.pbxproj (hunk 17 above)
Comment (karolkozub) [inline GrandCentralBoard.xcodeproj/project.pbxproj]: It seems that most extension files use the Objective-C category naming pattern, i.e. Class+Category.extension

Comment (ochococo) [inline GrandCentralBoard.xcodeproj/project.pbxproj]: In Swift? I don't think so.

File: GrandCentralBoard/Widgets/Harvest/View/AreaBarChartComponentViewModel+FromDailyBillingStats.swift (hunk 16 above)
Comment (ochococo) [inline GrandCentralBoard/Widgets/Harvest/View/AreaBarChartComponentViewModel+FromDailyBillingStats.swift]: Then you should also have opening and closing bracket on the same indentation level and so on...

**Comment by ochococo** [comment]
Awesome job @karolkozub 👏

**Pull request merged**